lib/options_validator.js

```diff
@@ -71,18 +94,39 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
 
   const verifiedOptions = Object.assign({}, providedOptions);
 
-  if (validationLevel === VALIDATION_LEVEL_NONE && overrideOptions == null) {
-    return Object.freeze(verifiedOptions);
+  if (
+    validationLevel === VALIDATION_LEVEL_NONE &&
+    overrideOptions == null &&
+    optionDefaults == null
+  ) {
+    return verifiedOptions;
   }
 
   Object.keys(optionsSchema).forEach(optionName => {
     const optionFromSchema = optionsSchema[optionName];
 
-    if (overrideOptions && overrideOptions[optionName] != null) {
+    if (
+      optionFromSchema.hasOwnProperty('overrideOnly') &&
+      verifiedOptions.hasOwnProperty(optionName)
+    ) {
+      console.warn(
+        `A value was provided for option [${optionName}], but this option cannot be specified. It will be overwritten.`
+      );
+    }
+
+    if (optionDefaults && optionDefaults.hasOwnProperty(optionName)) {
       if (!verifiedOptions.hasOwnProperty(optionName)) {
-        verifiedOptions[optionName] = overrideOptions[optionName];
+        verifiedOptions[optionName] = optionDefaults[optionName];
       }
-      if (optionFromSchema && optionFromSchema.default) {
+    }
+
+    if (overrideOptions && overrideOptions.hasOwnProperty(optionName)) {
+      verifiedOptions[optionName] = overrideOptions[optionName];
+
+      if (
+        (optionFromSchema && optionFromSchema.hasOwnProperty('default')) ||
+        (optionDefaults && optionDefaults.hasOwnProperty(optionName))
+      ) {
         console.warn(
           `A default value and override value were provided for option [${optionName}]. The override value will be used.`
         );
```

Comment:
In adding the documentation for `validate()`, I realized that I am only [setting non-dynamic defaults](https://github.com/mongodb/node-mongodb-native/pull/1859/files#diff-e1fa59f7e4e4c9080c24be96453dfdb8R142) if `validationLevel !== VALIDATION_LEVEL_NONE`. It seems like I should be setting non-dynamic defaults regardless of the validation level. Also, deprecation warnings are only being emitted if the `validationLevel` is `warn` or `error`. I think that we should be emitting deprecation warnings regardless of validation level. It seems to me that the only thing the `validationLevel` should affect is type validation. What do you both think?

Reply:
I agree. Validation level should only affect type validation. Defaults should always be set, and deprecation warnings should always be emitted (and then handled by their respective node handlers)

Reply:
Also agree, can you please add some tests validating this behavior when you change it.